frontend/src/Reducer-context.js

```diff
@@ -0,0 +1,313 @@
+      </Reducercontext.Provider>
+    </>
+  );
+}
+
+export function useReduce() {
+  return useContext(Reducercontext);
+}
+
+function reduce(state, action) {
+  switch (action.type) {
+    case "ROUTING":
+      return { ...state, route: action.payload };
+
+    case "PLAYING":
+      return { ...state, videoobj: action.payload };
+
+    case "SORT":
+      return { ...state, sortBy: action.payload };
+
+    case "FILTERCATEGORY":
+      if (state.showCategory.includes(action.payload)) {
+        return {
+          ...state,
+          showCategory: state.showCategory.filter(
+            (item) => item !== action.payload
+          )
+        };
+      } else {
+        return {
+          ...state,
+          showCategory: [...state.showCategory, action.payload]
+        };
+      }
+    case "CLEAR_FILTER":
+      return {
+        ...state,
+        showCategory: []
+      };
+
+    case "NEW_PLAYLIST":
+      let { id, name } = action.payload;
+      if (
+        state.playlist.find((item) => item.id === id || item.name === name) !==
+        undefined
+      ) {
+        return { ...state };
+      }
+      return {
+        ...state,
+        playlist: [...state.playlist, action.payload]
+      };
+
+    case "REMOVE_PLAYLIST":
+      return {
+        ...state,
+        playlist: state.playlist.filter((item) => action.payload !== item.id)
+      };
+
+    case "ADD_TO_PLAYLIST": {
+      let { playlistid, videoid } = action.payload;
+
+      return {
+        ...state,
+        playlist: state.playlist.map((item) => {
+          if (item.id === playlistid) {
+            return {
+              id: item.id,
+              name: item.name,
+              videos:
+                item.videos.find((item1) => item1 === videoid) === undefined
+                  ? item.videos.concat(videoid)
+                  : item.videos
+            };
+          }
+          return item;
+        })
+      };
+    }
+
+    case "REMOVE_FROM_PLAYLIST": {
+      let { playlistid, videoid } = action.payload;
+      console.log("reducer", playlistid, videoid);
+      return {
+        ...state,
+        playlist: state.playlist.map((item) => {
+          if (item.id === playlistid) {
+            return {
+              id: item.id,
+              name: item.name,
+              videos: item.videos.filter((item1) => item1 !== videoid)
+            };
+          }
+          return item;
+        })
+      };
+    }
+
+    case "ADD_TO_HISTORY":
+      let historyflag = 0;
+      let historynewarr = state.history.map((item) => {
+        if (item.id === action.payload.id) {
+          historyflag = 1;
+          return {
+            ...item,
+            ishistory: action.payload.ishistory,
+            lastseen: action.payload.lastseen
+          };
+        }
+        return item;
+      });
+      return historyflag === 0
+        ? { ...state, history: [...state.history, action.payload] }
+        : { ...state, history: historynewarr };
```

Comment:
Instead of using a flag variable, you can use the find method to do the same operation.